dali/kernels/imgproc/convolution/filter_gpu.cuh

```diff
@@ -0,0 +1,1094 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_CONVOLUTION_FILTER_GPU_CUH_
+#define DALI_KERNELS_IMGPROC_CONVOLUTION_FILTER_GPU_CUH_
+
+#include <limits>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "dali/core/boundary.h"
+#include "dali/core/convert.h"
+#include "dali/core/cuda_rt_utils.h"
+#include "dali/core/geom/vec.h"
+#include "dali/core/span.h"
+#include "dali/core/static_switch.h"
+#include "dali/core/tensor_view.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/imgproc/roi.h"
+#include "dali/kernels/kernel.h"
+
+namespace dali {
+namespace kernels {
+namespace filter {
+
+using boundary::BoundaryType;
+
+template <int N, typename T>
+vec<N, T> rev(const vec<N, T>& v) {
+  vec<N, T> out;
+  for (int d = 0; d < N; d++) {
+    out[N - d - 1] = v[d];
+  }
+  return out;
+}
+
+template <int N, typename T, typename U>
+void strides(vec<N, U>& out, U& total_stride, const vec<N, T>& v) {
+  total_stride = 1;
+  for (int d = 0; d < N; d++) {
+    out[d] = total_stride;
+    total_stride *= v[d];
+  }
+}
+
+template <int N, typename T, typename U>
+void strides(vec<N, U>& out, const vec<N, T>& v) {
+  U total_strides;
+  strides(out, total_strides, v);
+}
+
+template <int axes_>
+struct ShapeDesc {
+  static constexpr int axes = axes_;
+
```

Comment:
Alternatively, you could hold `Surface<axes, T>` instances, that would abstract away the multiplication by strides, etc.